lib/rules/array-type.js

```diff
@@ -0,0 +1,234 @@
+ * @returns {*} true or false
+ */
+function isSimpleType(node) {
+    switch (node.type) {
+        case "Identifier":
+        case "TSAnyKeyword":
+        case "TSBooleanKeyword":
+        case "TSNeverKeyword":
+        case "TSNumberKeyword":
+        case "TSObjectKeyword":
+        case "TSStringKeyword":
+        case "TSSymbolKeyword":
+        case "TSUnknownKeyword":
+        case "TSVoidKeyword":
+        case "TSNullKeyword":
+        case "TSArrayType":
+        case "TSUndefinedKeyword":
+        case "TSThisType":
+        case "TSQualifiedName":
+            return true;
+        case "TSTypeReference":
+            if (
+                node.typeName &&
+                node.typeName.type === "Identifier" &&
+                node.typeName.name === "Array"
+            ) {
+                if (!node.typeParameters) {
+                    return true;
+                }
+                if (node.typeParameters.params.length === 1) {
+                    return isSimpleType(node.typeParameters.params[0]);
+                }
+            } else {
+                if (node.typeParameters) {
+                    return false;
+                }
+                return isSimpleType(node.typeName);
+            }
+            return false;
+        default:
+            return false;
+    }
+}
+
+/**
+ * Check if node needs parentheses
+ * @param {ASTNode} node the node to be evaluated.
+ * @returns {*} true or false
+ */
+function typeNeedsParentheses(node) {
+    if (node.type === "TSTypeReference") {
+        switch (node.typeName.type) {
+            case "TSUnionType":
+            case "TSFunctionType":
+            case "TSIntersectionType":
+            case "TSTypeOperator":
+                return true;
+            default:
+                return false;
+        }
+    }
+    return false;
+}
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        docs: {
+            description: "Requires using either 'T[]' or 'Array' for arrays",
+            extraDescription: [util.tslintRule("array-type")],
+            category: "TypeScript",
+            url:
+                "https://github.com/nzakas/eslint-plugin-typescript/blob/master/docs/rules/array-type.md",
+        },
+        fixable: "code",
+        messages: {
+            errorStringGeneric:
+                "Array type using 'T[]' is forbidden. Use 'Array<T>' instead.",
+            errorStringGenericSimple:
+                "Array type using 'T[]' is forbidden for non-simple types. Use 'Array<T>' instead.",
+            errorStringArray:
+                "Array type using 'Array<T>' is forbidden. Use 'T[]' instead.",
+            errorStringArraySimple:
+                "Array type using 'Array<T>' is forbidden for simple types. Use 'T[]' instead.",
```

Comment:
To make the errors nicer, could you please make use of the `{{placeholders}}`?

Reply:
Mmm good catch. If they use a complex type then you'd have to make sure the message is correct too.

Can we handle the default case where it's just a single identifier (I.e. Use Foo from `Foo[]` or `Array<Foo>`), and for all other cases just use `T`?

That should give the majority case nice errors and save us from the ugly.

Reply:
ok, if it's simple type i will update message

Reply:
ok, i did change for that, in https://github.com/bradzacher/eslint-plugin-typescript/pull/211/commits/de4ca9d936fe20c73ffe6d4c66c19ec87c326520#diff-00f3c0ea3f6c5b8765c5c323ce639e24R126